### ISSUE TITLE
Blocked service registrations still run health checks

### ISSUE THREAD

**Issue by jessemyers** [hashicorp/consul#1263]
I've configured a Consul cluster with ACLs enabled and the default policy set to deny. I then created a service definition using the agent API _without_ passing an ACL token. The agent logs show that the service registration is blocked, but the service health checks still run.

IIUC, this means that unauthorized actors can cause Consul to run any executable it has access to.

I'm using Consul 0.5.2:

```
$ consul --version
Consul v0.5.2
Consul Protocol: 2 (Understands back to: 1)
```

The relevant portions of the server configurations are as follows:

```
"acl_datacenter": "dc1",
"acl_master_token": "secret",
"acl_default_policy": "deny",
"acl_down_policy": "deny",
```

I'm creating the check using:

```
http PUT localhost:8500/v1/agent/service/register Name=test Check:='{"Script": "/path/to/test.sh", "Interval": "15s"}'
```

where test.sh is:

```
#!/bin/bash

echo "OK" >> /tmp/out
```

Only the master and anonymous tokens exist. The anonymous token has no rules (and should deny all). 

I can see the following in the logs:

```
agent: Check 'service:test' registration blocked by ACLs
```

However, tailing /tmp/out still shows checks running.

Is this expected? Can it be prevented?

**Comment by ryanuber** [comment]
Hey Jesse,

This is currently expected behavior. Consul agents maintain their own local checks and services, and attempt to sync them with the servers periodically. The ACL enforcement is currently only done on the server side, so what's happening here is you have a working agent-side check which you can see is running periodically, but with an invalid token, preventing it from syncing to the servers, as you found in the logs. I'm marking this as a thinking ticket, since it would be nice to reject at the agent, but this would probably need to be a new opt-in argument as it affects other parts of the system.

**Comment by jessemyers** [comment]
Thanks for the response.

We're wrestling a bit with the power of Consul checks. If we allow Consul
agents to run (nearly) arbitrary commands for health checks, we make it
very easy for authors of individual services to be responsible for their
health checks. On the other hand, this creates a large attack surface for
malicious commands. We'd hoped to use ACLs to block check registrations
except for authorized processes (instead of spending more time locking down
what Consul can run.)

**Comment by slackpad** [comment]
This was fixed in Consul 0.8.0 where we now preflight ACL checks on the agent side. The service/checks will not be able to be registered w/o a token with sufficient access.
